### PR TITLE
leveldb: adjust compaction parameters based on usage

### PR DESCRIPTION
**Pull request by jakob223**
This spends less time compacting... but it spends a lot more time on the mutexes and timers involved in timing out when to trigger. Working on reducing that.

### CODE REVIEW

**Comment by jakob223** [comment]
I commented about the performance of this on the JIRA ticket, KBFS-3928, but there is one unresolved issue - the data race above. I fear that adding a lock on the options could make the LevelDB instance slower by a larger amount than that by which this is an improvement, but other options for changing the options live seem rather invasive. Thoughts?

**Comment by jzila** [comment]
@jakob223 even a RWMutex?

File: go/kbfs/libkbfs/leveldb.go [hunk 1]
```diff
@@ -35,19 +43,87 @@ var leveldbOptions = &opt.Options{
...
+	}
 }
 
 // Close closes the DB.
 func (ldb *LevelDb) Close() (err error) {
 	err = ldb.DB.Close()
 	// Hide the closer error.
-	_ = ldb.closer.Close()
+	_ = ldb.storageCloser.Close()
+	ldb.shutdownC <- struct{}{}
```
Comment (jzila) [inline go/kbfs/libkbfs/leveldb.go]: This is usually done as a `close(ldb.shutdownCh)`

File: go/kbfs/libkbfs/leveldb.go [hunk 2]
```diff
@@ -35,19 +43,87 @@ var leveldbOptions = &opt.Options{
...
 	*leveldb.DB
-	closer io.Closer
+	storageCloser io.Closer
+	options       *opt.Options
+
+	compactionLock   sync.Mutex
+	compactionQueued bool
+	compactionTime   time.Time
+	compactionTimer  *time.Timer
+	shutdownC        chan struct{}
```
Comment (jzila) [inline go/kbfs/libkbfs/leveldb.go]: We usually use `Ch` as our channel namer (e.g. `shutdownCh`).

File: go/kbfs/libkbfs/leveldb.go [hunk 3]
```diff
@@ -35,19 +43,87 @@ var leveldbOptions = &opt.Options{
...
+}
+
+// resets the compaction timer. When the compaction timer hits 0,
+// the DB considers compaction more strongly.
+func (ldb *LevelDb) resetTimer() {
+	ldb.compactionLock.Lock()
+	defer ldb.compactionLock.Unlock()
+	ldb.compactionTime = time.Now().Add(compactionTimeout)
```
Comment (jzila) [inline go/kbfs/libkbfs/leveldb.go]: Should probably be using a `clockwork.Clock` (https://godoc.org/github.com/jonboulle/clockwork#Clock), so you can simulate pauses in tests.

File: go/kbfs/libkbfs/leveldb.go [hunk 4]
```diff
@@ -35,19 +43,87 @@ var leveldbOptions = &opt.Options{
...
+
+	// set options to more compaction friendly
+	ldb.options.CompactionL0Trigger /= 2
+	ldb.options.CompactionTotalSize /= 2
+	ldb.DB.SetOptions(ldb.options)
+	ldb.options.CompactionL0Trigger *= 2
+	ldb.options.CompactionTotalSize *= 2
+
+	// make a trivial write to trigger compaction
+	err := ldb.DB.Put(compactionKey, compactionValue, nil)
```
Comment (jzila) [inline go/kbfs/libkbfs/leveldb.go]: This is dangerous, since it'll conflict with eviction.

**Comment by jakob223** [comment]
As we discussed, I'm going to hold off on doing more work here until after Journaling V3 has made more progress. However, the most recent test results I posted in the ticket are not promising :/

**Comment by jakob223** [comment]
Closing this until and unless we have reason to believe it will be useful.

**Pull request closed without merging**